### PR TITLE
performance hacking for minimize-regs

### PR DESCRIPTION
**Pull request by rachitnigam**

### CODE REVIEW

File: calyx/src/passes/minimize_regs.rs [hunk 1]
```diff
@@ -54,62 +40,70 @@ impl Visitor for MinimizeRegs {
...
+                acc.extend(live.get(&conf_group));
+                acc
+            })
+            .into_iter()
+            .for_each(|(group, confs)| {
+                let conflicts = live.get(&group);
+                confs
+                    .into_iter()
+                    // This unique call saves a lot of time!
+                    .unique()
```
Comment (rachitnigam) [inline calyx/src/passes/minimize_regs.rs:82]: This call is really beneficial. I wonder why we even have so many duplicates in the first place.

File: calyx/src/passes/minimize_regs.rs [hunk 2]
```diff
@@ -54,62 +40,70 @@ impl Visitor for MinimizeRegs {
...
+                    })
+            });
 
-                if a.get_paramter(&"width".into())
-                    != b.get_paramter(&"width".into())
-                {
-                    self.graph.insert_conflict(a.name.clone(), b.name.clone());
+        // add constraints so that registers of different sizes can't be shared
+        registers
+            .tuple_combinations()
```
Comment (rachitnigam) [inline calyx/src/passes/minimize_regs.rs:97]: Filter first and then iterate on smaller vecs!

File: calyx/src/utils/weight_graph.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
+use itertools::Itertools;
```
Comment (rachitnigam) [inline calyx/src/utils/weight_graph.rs:1]: This pattern of storing indices keeps showing up. Make it a utility!

File: calyx/src/passes/minimize_regs.rs [hunk 4]
```diff
@@ -54,62 +40,70 @@ impl Visitor for MinimizeRegs {
...
-            let in_parallel: Vec<_> = self
-                .par_conflicts
-                .all_conflicts(&group)
-                .into_iter()
-                .flat_map(|group| {
-                    self.live
-                        .get(&*group.borrow())
-                        .iter()
-                        .cloned()
-                        .collect::<Vec<_>>()
```
Comment (rachitnigam) [inline calyx/src/passes/minimize_regs.rs:76]: Collect calls are remarkably expensive!

File: calyx/src/passes/minimize_regs.rs [hunk 5]
```diff
@@ -54,62 +40,70 @@ impl Visitor for MinimizeRegs {
...
+            })
+            .into_iter()
+            .for_each(|(group, confs)| {
+                let conflicts = live.get(&group);
+                confs
+                    .into_iter()
+                    // This unique call saves a lot of time!
+                    .unique()
+                    .for_each(|par_conflict| {
+                        for conflict_here in conflicts {
```
Comment (rachitnigam) [inline calyx/src/passes/minimize_regs.rs:84]: This loop was and is the costliest part of this pass.

File: calyx/src/analysis/schedule_conflicts.rs [hunk 6]
```diff
@@ -1,52 +1,69 @@
...
 #[derive(Default)]
 /// A conflict graph that describes which groups are being run in parallel
 /// to each other.
 pub struct ScheduleConflicts {
-    /// The conflict graph. Two groups have an edge between them if they
-    /// may run in parallel.
-    conflicts: Graph<GroupNode, ()>,
-    /// Map from the name of the group to its NodeIndex in the conflicts
-    /// graph.
-    index_map: HashMap<ir::Id, NodeIndex>,
```
Comment (rachitnigam) [inline calyx/src/analysis/schedule_conflicts.rs:18]: Weird interface fixed: Keys are `ir::Id` but nodes are `Group`s?

**Comment by rachitnigam** [comment]
Lots of small fixes and with inline comment explanations. 

10x10 systolic array takes 3 seconds on my machine after this. Originally, a 6x6 array would take over a 66 seconds.

**Comment by sampsyo** [comment]
That's amazing!!

**Pull request merged**